File: sol/solution/0101-0200/0102/Solution.erl
%% Definition for a binary tree node.
%%
%% -record(tree_node, {val = 0 :: integer(),
%%                     left = null  :: 'null' | #tree_node{},
%%                     right = null :: 'null' | #tree_node{}}).
level_order(Root) ->
    level_order_queue([Root], []).
level_order_queue([], Acc) ->
    lists:reverse(Acc);
level_order_queue(Q, Acc) ->
    Levels = [N || N <- Q, N =/= null],
    case Levels of
        [] -> lists:reverse(Acc);
        _ ->
            LevelVals = [N#tree_node.val || N <- Levels],
            Next = lists:foldl(fun(N, S) ->
                        lists:append(S, [N#tree_node.left, N#tree_node.right])
                    end, [], Levels),
            level_order_queue(Next, [LevelVals|Acc])
    end.

File: sol/solution/0101-0200/0102/Solution.ex
# Definition for a binary tree node.
#
# defmodule TreeNode do
#   @type t :: %__MODULE__{
#           val: integer,
#           left: TreeNode.t() | nil,
#           right: TreeNode.t() | nil
#         }
#   defstruct val: 0, left: nil, right: nil
# end
defmodule Solution do
  def level_order(nil), do: []
  def level_order(root) do
    level_order([root], [])
  end
  defp level_order([], acc), do: Enum.reverse(acc)
  defp level_order(nodes, acc) do
    values = Enum.map(nodes, & &1.val)
    next =
      nodes
      |> Enum.flat_map(fn node -> [node.left, node.right] end)
      |> Enum.filter(& &1)
    level_order(next, [values | acc])
  end
end
